DataFormats/TrackReco/interface/TrackExtraBase.h

```diff
@@ -44,6 +44,8 @@ class TrackExtraBase
         return m_nHits;
     }
 
+    /// accessor to RecHits
+    auto recHits() const { return TrackingRecHitRange(recHitsBegin(), recHitsEnd()); }

```

Comment:
Alternatively, you could just use `recHitsProduct()` that returns const reference to the actual container.

Reply:
I think it's not the same: https://github.com/cms-sw/cmssw/blob/master/DataFormats/TrackReco/interface/TrackExtraBase.h#L50

Begin would be `recHitsProduct().data().begin()+firstRecHit();` and end `recHitsBegin()+recHitsSize();`. So it seems to me that the desired range is just a part (sub-range of elements next to each other) of the original product. Unless `m_firstHit` and `m_nHits` would always  be zero and the size of the collection of course, I have not checked that.

Reply:
Ok, so we can leave out the `data()` call. Bit in the end it's not about the `data()`, it's about the `+firstRecHit()`, right? Iterating over `OwnVector` would iterate over the same elements as iterating over the base vector I think, not the window that is defined by `firstRecHit()` and `recHitsSize()`. So the little helper class is still needed.

Reply:
Whoops, you're right, I completely missed `+firstRecHit()` (and forgot how the RecHits are organized within `TrackExtra`, shame on me). That indeed implies `recHitsProduct()` does not work, and one does need a range wrapper to be able to use range-for.